File: src/bumperbot_py_examples/bumperbot_py_examples/simple_service_client.py
import rclpy
from rclpy.node import Node
from bumperbot_msgs.srv import AddTwoInts
import sys

class SimpleServiceClient(Node):
    def __init__(self, a , b):
        super().__init__('simple_service_client')
        self.client_ = self.create_client(AddTwoInts, 'add_two_ints')
        while not self.client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Service not available, waiting again...')
        self.request = AddTwoInts.Request()
        self.request.a = a
        self.request.b = b

        self.future_ = self.client_.call_async(self.request)

        self.future_.add_done_callback(self.responseCallback)

    def responseCallback(self, future):
        self.get_logger().info(f"Serive response add_two_ints: {future.result().sum}")



def main(args=None):
    rclpy.init(args=args)
    # The first arf is the name of the node
    if len(sys.argv) != 3:
        print("Wrong number of args!" \
        "Usage: ros2 run bumperbot_py_examples simple_service_client a b")
        return -1
    a = int(sys.argv[1])
    b = int(sys.argv[2])
    simple_service_client = SimpleServiceClient(a, b)
    rclpy.spin(simple_service_client)
    simple_service_client.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_service_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <bumperbot_msgs/srv/add_two_ints.hpp>

using namespace std::placeholders;
using namespace std::chrono_literals;


class SimpleServiceClient: public rclcpp::Node
{
public:
    SimpleServiceClient(int a, int b): Node("simple_service_client")
    {
        client_ = create_client<bumperbot_msgs::srv::AddTwoInts>("add_two_ints");
        auto request_ = std::make_shared<bumperbot_msgs::srv::AddTwoInts::Request>();
        request_->a = a;
        request_->b = b;

        while (!client_->wait_for_service(1s)) {
            if(!rclcpp::ok()) {
                RCLCPP_ERROR_STREAM(get_logger(), "Interrupted while waiting for the service. Exiting.");
                return;
            }
            RCLCPP_INFO_STREAM(get_logger(), "Service not available, waiting again...");
        }

        auto result_future = client_->async_send_request(request_,
            std::bind(&SimpleServiceClient::responseCallback, this, _1));
    }


private:
    rclcpp::Client<bumperbot_msgs::srv::AddTwoInts>::SharedPtr client_;

    void responseCallback(rclcpp::Client<bumperbot_msgs::srv::AddTwoInts>::SharedFuture future)
    {
        if (future.valid()) {
        RCLCPP_INFO_STREAM(get_logger(), "Response received: " << future.get()->sum);
        } else {
            RCLCPP_ERROR_STREAM(get_logger(), "Service call failed");
        }
    }
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc,argv);

    if (argc != 3) {
        RCLCPP_ERROR_STREAM(rclcpp::get_logger("rclcpp"), "Usage: simple_service_client <int a> <int b>");
        return 1;
    }

    int a = std::stoi(argv[1]);
    int b = std::stoi(argv[2]);

    auto node = std::make_shared<SimpleServiceClient>(a,b);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}